internal/models/sessions.go

```diff
@@ -69,13 +69,41 @@ type Session struct {
 	FactorID  *uuid.UUID `json:"factor_id" db:"factor_id"`
 	AMRClaims []AMRClaim `json:"amr,omitempty" has_many:"amr_claims"`
 	AAL       *string    `json:"aal" db:"aal"`
+
+	RefreshedAt *time.Time `json:"refreshed_at,omitempty" db:"refreshed_at"`
+	UserAgent   *string    `json:"user_agent,omitempty" db:"user_agent"`
+	IP          *string    `json:"ip,omitempty" db:"ip"`
 }
 
 func (Session) TableName() string {
 	tableName := "sessions"
 	return tableName
 }
 
+func (s *Session) LastRefreshedAt(refreshTokenTime *time.Time) time.Time {
+	refreshedAt := s.RefreshedAt
+
+	if refreshedAt == nil || refreshedAt.IsZero() {
+		if refreshTokenTime != nil {
+			rtt := *refreshTokenTime
+
+			if rtt.IsZero() {
+				return s.CreatedAt
+			} else if rtt.After(s.CreatedAt) {
+				return rtt
+			}
+		}
+
+		return s.CreatedAt
+	}
+
+	return *refreshedAt
+}
```

Comment:
i think we should use the `session.CreatedAt` value instead

Reply:
I thought so too... but while I was testing it over the weekend realized this scenario is not covered and feels kinda bad to me:

1. Imagine a user logged in 2 weeks ago, the user used the app for a while and then just didn't come back.
2. The developer sees the new feature next week and sets an inactivity timeout of 1 month.
3. (So far only 3 weeks have passed from the user logging in.)
4. Because that user's session has `refreshed_at` which is null, when they come back after the user set the inactivity timeout, they're kicked out (even though they have 1 more week before being "inactive".)

The alternative was to backfill the `sessions` table with the best `refreshed_at` values we can derive from `refresh_tokens`... but after spending all day trying to design a very efficient query it appears that's not possible (partly because of the need for a `GROUPED BY` clause in `refresh_tokens`). On supabase.com that would result in about 30s of backfilling which is fairly slow.

So I decided that it's not necessary to backfill, we just need to guesstimate based on the token's `updated_at` and that should cover most valid use cases.